// @sap/cds-dk/lib/init/template/dynatrace/index.js
const { merge, sort } = require('../../merge')

const { readProject } = require('../../projectReader')
const { srv4, mtxSidecar4 } = require('../../registries/mta')

module.exports = class extends require('../../plugin') {

  static hasInProduction(env) {
    return env.requires?.telemetry?.kind === 'to-dynatrace'
  }

  async canRun() {
    const { hasMta, hasHelm, hasHelmUnifiedRuntime } = readProject()
    if ((hasHelmUnifiedRuntime || hasHelm) && !hasMta) throw `'cds add dynatrace' is not available for Kyma yet`
    return true
  }

  async run() {
    const project = readProject()
    await merge(__dirname, 'files/package.json.hbs').into('package.json', { with: project })
    await sort('package.json', 'dependencies')
  }

  async combine() {
    const project = readProject()
    const { hasMta, isJava, hasMultitenancy, srvPath } = project

    if (hasMta) {
      const dynatrace = {
        in: 'resources',
        where: {
          type: 'org.cloudfoundry.managed-service',
          'parameters.service': 'dynatrace'
        }
      }
      const srv = srv4(srvPath)
      const additions = [srv, dynatrace]
      const relationships = [{
        insert: [dynatrace, 'name'],
        into: [srv, 'requires', 'name']
      }]
      if (hasMultitenancy) {
        const mtxSidecar = mtxSidecar4(isJava ? 'mtx/sidecar' : 'gen/mtx/sidecar')
        additions.push(mtxSidecar)
        relationships.push({
          insert: [dynatrace, 'name'],
          into: [mtxSidecar, 'requires', 'name']
        })
      }
      await merge(__dirname, 'files/mta.yaml.hbs').into('mta.yaml', {
        with: project, additions, relationships
      })
    }
  }
}
